<a href="https://colab.research.google.com/github/Thitichaya2000/data_viz_2022/blob/main/Pandas101.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [28]:
import pandas as pd  # ใช้จัดการข้อมูลที่หน้าตาเป็นตาราง

.CSV = Comma Separated Values

แหล่งข้อมูลที่นำมาใช้ : https://data.go.th/

In [29]:
from google.colab import drive  
drive.mount('/content/drive')
# เชื่อม google drive เพื่อเอาข้อมูลมาใช้

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [30]:
import os  # จัดการไฟล์ต่างๆ

In [31]:
path = '/content/drive/My Drive/dataviz_2022_data'  # ชี้โฟลเดอร์ใน google drive

In [32]:
covid_file_path = os.path.join(path,'pm-20-jan-2021.csv')  # os.path ที่อยู่ไฟล์นี้ .join ชื่อไฟล์
print(covid_file_path)

/content/drive/My Drive/dataviz_2022_data/pm-20-jan-2021.csv


In [33]:
a = path + '/pm-18-jan-2021.csv' 
print(a)

/content/drive/My Drive/dataviz_2022_data/pm-18-jan-2021.csv


###load data to memory

In [34]:
data_covid = pd.read_csv(covid_file_path)  # pandas อ่าน csv
data_covid

No.  ...                                         risk
0          1  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
1          2  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
2          3  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
3          4  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
4          5  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
...      ...  ...                                          ...
12648  12649  ...                             State Quarantine
12649  12650  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12650  12651  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12651  12652  ...                                Cluster ระยอง
12652  12653  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[12653 rows x 10 columns]

In [35]:
data_covid.head()

No. announce_date  ... district_of_onset                              risk
0    1     1/12/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
1    2     1/17/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
2    3     1/22/2020  ...             เมือง  คนต่างชาติเดินทางมาจากต่างประเทศ
3    4     1/22/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
4    5     1/24/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ

[5 rows x 10 columns]

In [36]:
data_covid.shape

(12653, 10)

###ชี้ค่าในตาราง

basic

In [37]:
data_covid['province_of_onset']

0        กรุงเทพมหานคร
1        กรุงเทพมหานคร
2               นครปฐม
3        กรุงเทพมหานคร
4        กรุงเทพมหานคร
             ...      
12648           ชลบุรี
12649            ระยอง
12650            ระยอง
12651            ระยอง
12652              ตาก
Name: province_of_onset, Length: 12653, dtype: object

In [38]:
data_covid['province_of_onset'][4]

'กรุงเทพมหานคร'

.iloc


In [39]:
data_covid.iloc[4,7]

'กรุงเทพมหานคร'

##Table slicing

###ตัดมาเฉพาะ column ที่ต้องการ

In [40]:
smaller_table = data_covid[['announce_date','province_of_onset','risk']]
smaller_table

announce_date  ...                                         risk
0         1/12/2020  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
1         1/17/2020  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
2         1/22/2020  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
3         1/22/2020  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
4         1/24/2020  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
...             ...  ...                                          ...
12648     1/20/2021  ...                             State Quarantine
12649     1/20/2021  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12650     1/20/2021  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12651     1/20/2021  ...                                Cluster ระยอง
12652     1/20/2021  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[12653 rows x 3 columns]

###ตัดเฉพาะ rows ที่ต้องการ

แบบง่าย แต่ใช้การอะไรไม่ค่อยจะได้ (มองเป็น array แล้วตัดแบบ array)

In [41]:
data_covid.iloc[1:5,:]

No. announce_date  ... district_of_onset                              risk
1    2     1/17/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
2    3     1/22/2020  ...             เมือง  คนต่างชาติเดินทางมาจากต่างประเทศ
3    4     1/22/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
4    5     1/24/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ

[4 rows x 10 columns]

แบบ advanced แต่ใช้ประโยชน์ได้เยอะกว่า (ตัดแบบใช้ logic query)

In [42]:
data_covid[data_covid['province_of_onset']=='ขอนแก่น']

No.  ...                                           risk
180      181  ...                                        สนามมวย
462      463  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
1466    1467  ...                  คนไทยเดินทางกลับจากต่างประเทศ
1970    1971  ...  ไปสถานที่ชุมชน เช่น ตลาดนัด สถานที่ท่องเที่ยว
2637    2638  ...                  สัมผัสผู้เดินทางจากต่างประเทศ
2673    2674  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
5948    5949  ...                              Cluster สมุทรสาคร
6082    6083  ...                              Cluster สมุทรสาคร
9333    9334  ...                           อยู่ระหว่างการสอบสวน
9334    9335  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
10610  10611  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11517  11518  ...                           อยู่ระหว่างการสอบสวน
11697  11698  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11698  11699  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[14 rows x 10 columns]

การชี้ค่าในตารางด้วย data_covid[x] ค่า x สามารถมีค่าที่ใส่เข้าไปได้ 2 แบบ

1.   ชื่อ column

2.   ใส่ rows ที่ต้องการ (ใส่ list ที่มีขนาดเท่ากับจำนวนแถว แล้วบอกว่าเอา แถวไหนบ้างโดยใช้ logical expression (True/False))

In [45]:
eight_rows_covid = data_covid.iloc[:8,:] # ตารางตัวอย่าง
eight_rows_covid

No. announce_date  ... district_of_onset                              risk
0    1     1/12/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
1    2     1/17/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
2    3     1/22/2020  ...             เมือง  คนต่างชาติเดินทางมาจากต่างประเทศ
3    4     1/22/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
4    5     1/24/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
5    6     1/25/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
6    7     1/26/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
7    8     1/26/2020  ...            หัวหิน  คนต่างชาติเดินทางมาจากต่างประเทศ

[8 rows x 10 columns]

In [46]:
eight_rows_covid[[True,False,True,True,False,True,False,False]]   #2.   ใส่ rows ที่ต้องการ

No. announce_date  ... district_of_onset                              risk
0    1     1/12/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
2    3     1/22/2020  ...             เมือง  คนต่างชาติเดินทางมาจากต่างประเทศ
3    4     1/22/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
5    6     1/25/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ

[4 rows x 10 columns]

In [47]:
eight_rows_covid['province_of_onset'] == 'กรุงเทพมหานคร'  #สร้าง list ของ logical expression

0     True
1     True
2    False
3     True
4     True
5     True
6     True
7    False
Name: province_of_onset, dtype: bool

In [48]:
eight_rows_covid[eight_rows_covid['province_of_onset'] == 'กรุงเทพมหานคร']  #เราเอา 2 บรรทัดบนมาบวกกัน

No. announce_date  ... district_of_onset                              risk
0    1     1/12/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
1    2     1/17/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
3    4     1/22/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
4    5     1/24/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
5    6     1/25/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
6    7     1/26/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ

[6 rows x 10 columns]

HW (ให้ใช้เฉพาะที่อาจารย์สอนไปแล้วในวิชานี้)

1. คำนวณ อายุเฉลี่ย ของผู้หญิง และผู้ชาย ของข้อมูลทั้งหมด
2. คำนวณ อายุเฉลี่ย ของผู้หญิง และผู้ชาย ของผู้ป่วยในจังหวัดขอนแก่น


หาจำนวนผู้ป่วยที่เป็นคน "คนต่างชาติเดินทางมาจากต่างประเทศ"

In [49]:
this_data = data_covid[['sex','age','province_of_onset']]

In [50]:
this_data

sex   age province_of_onset
0      หญิง  61.0     กรุงเทพมหานคร
1      หญิง  74.0     กรุงเทพมหานคร
2      หญิง  73.0            นครปฐม
3       ชาย  68.0     กรุงเทพมหานคร
4      หญิง  66.0     กรุงเทพมหานคร
...     ...   ...               ...
12648  หญิง  44.0            ชลบุรี
12649  หญิง  52.0             ระยอง
12650  หญิง  23.0             ระยอง
12651  หญิง  29.0             ระยอง
12652  หญิง  22.0               ตาก

[12653 rows x 3 columns]

In [51]:
female = this_data[this_data['sex']=='หญิง']
female['age']

0        61.0
1        74.0
2        73.0
4        66.0
5        33.0
         ... 
12648    44.0
12649    52.0
12650    23.0
12651    29.0
12652    22.0
Name: age, Length: 5332, dtype: float64

In [52]:
sum = 0
N = 0
for a in female['age']:
    
    if a > 0:
        sum += a # sum = sum + a
        N += 1

print(f'อายุเฉลี่ย ของ ผู้ป่วยหญิง {sum/N}')

อายุเฉลี่ย ของ ผู้ป่วยหญิง 36.63955450022614


In [53]:
female_KK = female[female['province_of_onset']=='ขอนแก่น']

sum = 0
N = 0
for a in female_KK['age']:
    
    if a > 0:
        sum += a # sum = sum + a
        N += 1

print(f'อายุเฉลี่ย ของ ผู้ป่วยหญิง ในจังหวัดขอนแก่น {sum/N}')

อายุเฉลี่ย ของ ผู้ป่วยหญิง ในจังหวัดขอนแก่น 41.57142857142857


อายุเฉลี่ย ของ ผู้ป่วยหญิง ในจังหวัดขอนแก่น 41.57142857142857

การจัดการ Missing Value

* ลบ record ที่เป็น missing
* แทนที่ ค่า missing ด้วยค่าที่เหมาะสม mean, default, category-unknown
* ใช้ ค่าจาก columns อื่นๆ ช่วยประมาณค่า ค่าใน column ที่หายไป (regression, deep learning, etc.)

###ลบ record (dropna)

missing = None NA(not autorized) NaN (not a number)

In [54]:
this_data.shape #(12423, 3)

(12653, 3)

In [55]:
print(data_covid.shape)
print(data_covid.dropna().shape)

(12653, 10)
(6655, 10)


In [56]:
this_data.dropna().shape

(8478, 3)

In [57]:
this_data.dropna()

sex   age province_of_onset
0      หญิง  61.0     กรุงเทพมหานคร
1      หญิง  74.0     กรุงเทพมหานคร
2      หญิง  73.0            นครปฐม
3       ชาย  68.0     กรุงเทพมหานคร
4      หญิง  66.0     กรุงเทพมหานคร
...     ...   ...               ...
12648  หญิง  44.0            ชลบุรี
12649  หญิง  52.0             ระยอง
12650  หญิง  23.0             ระยอง
12651  หญิง  29.0             ระยอง
12652  หญิง  22.0               ตาก

[8478 rows x 3 columns]

In [58]:
this_data.dropna()
print(this_data.shape)

(12653, 3)


In [59]:
this_data.dropna(inplace=True)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [ ]:
this_data.dropna(inplace=True)
print(this_data.shape)


In [60]:
this_data = data_covid[['sex','age','province_of_onset']]
this_data.shape

(12653, 3)

In [61]:
this_data.dropna().shape

(8478, 3)

In [62]:
this_data.dropna(subset=['age']).shape

(8881, 3)

###แทน missing ด้วยค่าที่เหมาะสม (fillna)

In [63]:
this_data = data_covid[['sex','age','province_of_onset']]

In [64]:
this_data.head()


sex   age province_of_onset
0  หญิง  61.0     กรุงเทพมหานคร
1  หญิง  74.0     กรุงเทพมหานคร
2  หญิง  73.0            นครปฐม
3   ชาย  68.0     กรุงเทพมหานคร
4  หญิง  66.0     กรุงเทพมหานคร

In [65]:
print(f'จำนวน record ก่อน drop missing ใน province {this_data.shape[0]}')
print(f'จำนวน record หลัง drop missing ใน province {this_data.dropna(subset=["province_of_onset"]).shape[0]}')

จำนวน record ก่อน drop missing ใน province 12653
จำนวน record หลัง drop missing ใน province 10662


In [66]:
this_data_updated = this_data.fillna(value={'sex':'ไม่รู้', 'age':-1, 'province_of_onset':'ไม่รู้'})
this_data_updated.head()

sex   age province_of_onset
0  หญิง  61.0     กรุงเทพมหานคร
1  หญิง  74.0     กรุงเทพมหานคร
2  หญิง  73.0            นครปฐม
3   ชาย  68.0     กรุงเทพมหานคร
4  หญิง  66.0     กรุงเทพมหานคร

In [67]:
this_data_updated[this_data_updated['sex']=='ไม่รู้']

sex  age province_of_onset
4391   ไม่รู้ -1.0         สมุทรสาคร
4392   ไม่รู้ -1.0         สมุทรสาคร
4393   ไม่รู้ -1.0         สมุทรสาคร
4394   ไม่รู้ -1.0         สมุทรสาคร
4395   ไม่รู้ -1.0         สมุทรสาคร
...       ...  ...               ...
12635  ไม่รู้ -1.0         สมุทรสาคร
12636  ไม่รู้ -1.0         สมุทรสาคร
12637  ไม่รู้ -1.0         สมุทรสาคร
12638  ไม่รู้ -1.0         สมุทรสาคร
12639  ไม่รู้ -1.0         สมุทรสาคร

[2502 rows x 3 columns]

In [68]:
this_data_updated['province_of_onset']=='ไม่รู้'

0        False
1        False
2        False
3        False
4        False
         ...  
12648    False
12649    False
12650    False
12651    False
12652    False
Name: province_of_onset, Length: 12653, dtype: bool

In [69]:
data_covid[this_data_updated['province_of_onset']=='ไม่รู้']

No.  ...                                         risk
414      415  ...                                      สนามมวย
529      530  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
555      556  ...                                      สนามมวย
576      577  ...                                      สนามมวย
588      589  ...                         อยู่ระหว่างการสอบสวน
...      ...  ...                                          ...
12405  12406  ...                            Cluster สมุทรสาคร
12406  12407  ...                            Cluster สมุทรสาคร
12407  12408  ...                            Cluster สมุทรสาคร
12408  12409  ...                            Cluster สมุทรสาคร
12409  12410  ...                            Cluster สมุทรสาคร

[1991 rows x 10 columns]

In [70]:
this_data_updated_2 = this_data.fillna(value={'province_of_onset':'ไม่รู้'})


In [71]:
this_data_updated_2[this_data_updated['sex']=='ไม่รู้']

sex  age province_of_onset
4391   NaN  NaN         สมุทรสาคร
4392   NaN  NaN         สมุทรสาคร
4393   NaN  NaN         สมุทรสาคร
4394   NaN  NaN         สมุทรสาคร
4395   NaN  NaN         สมุทรสาคร
...    ...  ...               ...
12635  NaN  NaN         สมุทรสาคร
12636  NaN  NaN         สมุทรสาคร
12637  NaN  NaN         สมุทรสาคร
12638  NaN  NaN         สมุทรสาคร
12639  NaN  NaN         สมุทรสาคร

[2502 rows x 3 columns]

###การวนลูป record ในตาราง (.iterrows)

In [72]:
this_data = data_covid[['sex','age','province_of_onset']]

In [73]:
for each_row in this_data.iterrows():
    # print(each_row)
    # print(each_row[1])
    if (each_row[1]['age'] == 20) and (each_row[1]['province_of_onset'] == 'ขอนแก่น'):
        print(each_row)

(9334, sex                      ชาย
age                       20
province_of_onset    ขอนแก่น
Name: 9334, dtype: object)


###การวนลูป แบบมองตารางแพนด้าส์(pandas dataframe) เป็น numpy array (.iloc)

In [74]:
for each_row in range(this_data.shape[0]):
    if (this_data.iloc[each_row,1] == 20) and (this_data.iloc[each_row,2] == 'ขอนแก่น'):  #each_row คือ ตัวเลข
        print(each_row)
        print(this_data.iloc[each_row,:])   

9334
sex                      ชาย
age                       20
province_of_onset    ขอนแก่น
Name: 9334, dtype: object


###Quiz ตัดตารางออกมาเฉพาะปี 2021 announce_date ในปี 2021

Hint

* วนลูปหา index ของปี 2021
* ตัดตารางมาเฉพาะ ปี 2021

In [75]:
TF=list()
for each_row in data_covid.iterrows():
    if each_row[1]['announce_date'].split('/')[2] == '2021':
        TF.append(True)
    else:
        TF.append(False)

data_covid[TF].head()

No.  ...                                               risk
6884  6885  ...                                  Cluster สมุทรสาคร
6885  6886  ...                                      Cluster ระยอง
6886  6887  ...                                        สถานบันเทิง
6887  6888  ...                                  Cluster สมุทรสาคร
6888  6889  ...  อาชีพเสี่ยง เช่น ทำงานในสถานที่แออัด หรือทำงาน...

[5 rows x 10 columns]

###Function ตัวช่วยใน pandas

.describe() คำนวณค่าทางสถิติของข้อมูลที่เป็นตัวเลข

In [76]:
data_covid.describe()

No.          age
count  12653.000000  8881.000000
mean    6327.000000    38.054748
std     3652.750813    15.125178
min        1.000000     0.100000
25%     3164.000000    27.000000
50%     6327.000000    36.000000
75%     9490.000000    48.000000
max    12653.000000    97.000000

.mean() คำนวณค่าเฉลี่ยของข้อมูลโดยไม่สนใจ missing

In [77]:
data_covid[data_covid['sex']=='ชาย']['age'].mean()

39.46076940331987

.isnull()

In [78]:
data_covid.isnull()

No.  announce_date  ...  district_of_onset   risk
0      False          False  ...               True  False
1      False          False  ...               True  False
2      False          False  ...              False  False
3      False          False  ...               True  False
4      False          False  ...               True  False
...      ...            ...  ...                ...    ...
12648  False          False  ...              False  False
12649  False          False  ...              False  False
12650  False          False  ...              False  False
12651  False          False  ...              False  False
12652  False          False  ...              False  False

[12653 rows x 10 columns]

###next
* .describe
* .mean
* HW วนลูป missing .isnull
* ต่อตารางแกน X แกน y
* .groupby
* save table